File: bot.ts
import "dotenv/config";
import { Bot, Context, InlineKeyboard, session, SessionFlavor } from "grammy";
import { getLast3Posts, validateUsername, Tweet } from "./twitterClient";
import { generateReplies } from "./replyGenerator";
import { addSub, getUserSubs, loadSubs, markSent, removeSub, wasSent } from "./subscriptions";

interface SessionData {
  waitingForGetPosts: boolean;
  waitingForSub: boolean;
}

type MyContext = Context & SessionFlavor<SessionData>;

const token = process.env.TELEGRAM_BOT_TOKEN;
if (!token) throw new Error("TELEGRAM_BOT_TOKEN is not set");

const bot = new Bot<MyContext>(token);
bot.use(session<SessionData, MyContext>({ initial: () => ({ waitingForGetPosts: false, waitingForSub: false }) }));

function parseUsername(raw: string): string {
  const m = raw.match(/(?:twitter\.com|x\.com)\/([A-Za-z0-9_]+)/);
  if (m) return m[1];
  return raw.trim().replace(/^@/, "");
}

function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function formatTweetMessage(prefix: string, tweet: Tweet, reply: string): string {
  const preview = tweet.text.length > 200 ? tweet.text.slice(0, 200) + "…" : tweet.text;
  return (
    `<b>${prefix}</b>\n${escapeHtml(preview)}\n\n` +
    `🔗 ${tweet.url}\n\n` +
    `💬 <b>Предлагаемый ответ:</b>\n${escapeHtml(reply)}`
  );
}

async function showSubscriptions(ctx: MyContext, edit = false) {
  const userId = ctx.from!.id;
  const subs = getUserSubs(userId);
  const text = subs.length
    ? `Твои подписки (${subs.length}):\n${subs.map((u) => `• @${u}`).join("\n")}`
    : "У тебя пока нет подписок.";

  const keyboard = new InlineKeyboard().text("➕ Добавить аккаунт", "sub_add");
  for (const u of subs) {
    keyboard.row().text(`❌ Убрать @${u}`, `sub_remove_${u}`);
  }

  const opts = { reply_markup: keyboard, parse_mode: "HTML" as const };
  if (edit) await ctx.editMessageText(text, opts);
  else await ctx.reply(text, opts);
}

async function fetchAndSendPosts(ctx: MyContext, username: string) {
  if (!username) {
    ctx.session.waitingForGetPosts = true;
    await ctx.reply("Не удалось распознать аккаунт. Напиши @username или ссылку на профиль X/Twitter:");
    return;
  }

  const statusMsg = await ctx.reply(`Ищу посты @${username}…`);
  const editStatus = (text: string) =>
    ctx.api.editMessageText(statusMsg.chat.id, statusMsg.message_id, text);

  let unliked: Tweet[], all: Tweet[];
  try {
    ({ unliked, all } = await getLast3Posts(username));
  } catch (e: any) {
    ctx.session.waitingForGetPosts = true;
    await editStatus(
      `Аккаунт @${username} не найден в X/Twitter.\n\nНапиши другой @username или ссылку на профиль:`
    );
    return;
  }

  if (!all.length) {
    await editStatus(`У @${username} нет постов.`);
    return;
  }

  if (!unliked.length) {
    await editStatus(`Все последние ${all.length} поста @${username} уже лайкнуты ✅`);
    return;
  }

  await ctx.api.deleteMessage(statusMsg.chat.id, statusMsg.message_id);

  const replies = await generateReplies(unliked);
  for (let i = 0; i < unliked.length; i++) {
    await ctx.reply(formatTweetMessage(`🐦 Пост @${username}:`, unliked[i], replies[i]), {
      parse_mode: "HTML",
    });
  }
}

// ── Commands ─────────────────────────────────────────────────────────────────

bot.command("start", async (ctx) => {
  await ctx.reply(
    "Привет!\n\n" +
    "/get_posts @username — последние 3 поста (нелайкнутые придут с ответами)\n" +
    "/subscriptions — подписки (проверяю нелайкнутые каждые 2 часа)"
  );
});

bot.command("get_posts", async (ctx) => {
  const arg = ctx.match.trim();
  if (!arg) {
    ctx.session.waitingForGetPosts = true;
    await ctx.reply("Напиши @username или ссылку на профиль X/Twitter:");
    return;
  }
  await fetchAndSendPosts(ctx, parseUsername(arg));
});

bot.command("subscriptions", async (ctx) => {
  await showSubscriptions(ctx);
});

// ── Callback queries ──────────────────────────────────────────────────────────

bot.callbackQuery("sub_add", async (ctx) => {
  await ctx.answerCallbackQuery();
  ctx.session.waitingForSub = true;
  await ctx.reply("Напиши @username или ссылку на аккаунт X/Twitter:");
});

bot.callbackQuery(/^sub_remove_(.+)$/, async (ctx) => {
  await ctx.answerCallbackQuery();
  removeSub(ctx.from.id, ctx.match[1]);
  await showSubscriptions(ctx, true);
});

// ── Text messages ─────────────────────────────────────────────────────────────

bot.on("message:text", async (ctx) => {
  if (ctx.session.waitingForGetPosts) {
    ctx.session.waitingForGetPosts = false;
    await fetchAndSendPosts(ctx, parseUsername(ctx.message.text));
    return;
  }

  if (ctx.session.waitingForSub) {
    ctx.session.waitingForSub = false;
    const username = parseUsername(ctx.message.text);
    if (!username) {
      await ctx.reply("Не удалось распознать имя аккаунта.");
      return;
    }
    if (addSub(ctx.from.id, username)) {
      await ctx.reply(`✅ @${username} добавлен в подписки`);
      await checkOneSubscription(ctx.from.id, username);
    } else {
      await ctx.reply(`@${username} уже в подписках`);
    }
    return;
  }

  await ctx.reply("Используй:\n/get_posts @username — посты\n/subscriptions — подписки");
});

// ── Subscription checks ───────────────────────────────────────────────────────

async function checkOneSubscription(userId: number, username: string) {
  try {
    const { unliked } = await getLast3Posts(username);
    const fresh = unliked.filter((t) => !wasSent(userId, username, t.id));
    if (!fresh.length) return;

    const replies = await generateReplies(fresh);
    for (let i = 0; i < fresh.length; i++) {
      await bot.api.sendMessage(
        userId,
        formatTweetMessage(`🔔 Новый пост @${username}:`, fresh[i], replies[i]),
        { parse_mode: "HTML" }
      );
      markSent(userId, username, fresh[i].id);
    }
  } catch (e) {
    console.error(`Subscription check failed for @${username} / user ${userId}:`, e);
  }
}

async function checkAllSubscriptions() {
  const data = loadSubs();
  for (const [userIdStr, usernames] of Object.entries(data)) {
    for (const username of usernames) {
      await checkOneSubscription(Number(userIdStr), username);
    }
  }
}

// ── Start ─────────────────────────────────────────────────────────────────────

async function main() {
  await bot.api.setMyCommands([
    { command: "get_posts", description: "Последние 3 поста аккаунта X/Twitter" },
    { command: "subscriptions", description: "Управление подписками" },
    { command: "start", description: "Начать работу с ботом" },
  ]);

  const me = await bot.api.getMe();
  console.log(`Бот запущен: @${me.username} (id=${me.id}). Ctrl+C для остановки.`);

  setInterval(async () => {
    console.log("Running subscription check…");
    await checkAllSubscriptions();
  }, 2 * 60 * 60 * 1000);

  await bot.start();
}

main().catch(console.error);

File: bot.py
import asyncio
import logging
import os
import re

from dotenv import load_dotenv
from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.ext import (
    Application,
    CallbackQueryHandler,
    CommandHandler,
    ContextTypes,
    MessageHandler,
    filters,
)

from reply_generator import generate_replies
from subscriptions import (
    add_sub,
    get_user_subs,
    load_subs,
    mark_sent,
    remove_sub,
    was_sent,
)
from twitter_client import get_last_3_posts

load_dotenv()
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(name)s: %(message)s",
)


def parse_username(raw: str) -> str:
    raw = raw.strip()
    m = re.search(r"(?:twitter\.com|x\.com)/([A-Za-z0-9_]+)", raw)
    if m:
        return m.group(1)
    return raw.lstrip("@")


def _format_tweet_message(prefix: str, username: str, tweet: dict, reply: str) -> str:
    preview = tweet["text"][:200] + ("…" if len(tweet["text"]) > 200 else "")
    return (
        f"{prefix}\n{preview}\n\n"
        f"🔗 {tweet['url']}\n\n"
        f"💬 *Предлагаемый ответ:*\n{reply}"
    )


async def _show_subscriptions(target, user_id: int, edit: bool = False):
    subs = get_user_subs(user_id)
    text = (
        f"Твои подписки ({len(subs)}):\n" + "\n".join(f"• @{u}" for u in subs)
        if subs
        else "У тебя пока нет подписок."
    )
    keyboard = [[InlineKeyboardButton("➕ Добавить аккаунт", callback_data="sub_add")]]
    for u in subs:
        keyboard.append([InlineKeyboardButton(f"❌ Убрать @{u}", callback_data=f"sub_remove_{u}")])
    markup = InlineKeyboardMarkup(keyboard)
    if edit:
        await target.edit_message_text(text, reply_markup=markup)
    else:
        await target.message.reply_text(text, reply_markup=markup)


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "Привет!\n\n"
        "/getPosts @username — последние 3 поста (нелайкнутые придут с ответами)\n"
        "/subscriptions — подписки (проверяю нелайкнутые каждые 2 часа)"
    )


async def _fetch_and_send_posts(update: Update, context: ContextTypes.DEFAULT_TYPE, username: str):
    """Fetch last 3 posts for username and send results. Sets waiting_for_getposts if profile not found."""
    if not username:
        context.user_data["waiting_for_getposts"] = True
        await update.message.reply_text(
            "Не удалось распознать аккаунт. Напиши @username или ссылку на профиль X/Twitter:"
        )
        return

    status = await update.message.reply_text(f"Ищу посты @{username}…")
    try:
        unliked, all_tweets = get_last_3_posts(username)
    except ValueError as e:
        context.user_data["waiting_for_getposts"] = True
        await status.edit_text(
            f"Аккаунт @{username} не найден в X/Twitter.\n\nНапиши другой @username или ссылку на профиль:"
        )
        return
    except Exception as e:
        await status.edit_text(f"Ошибка при получении твитов: {e}")
        return

    if not all_tweets:
        await status.edit_text(f"У @{username} нет постов.")
        return

    if not unliked:
        await status.edit_text(f"Все последние {len(all_tweets)} поста @{username} уже лайкнуты ✅")
        return

    await status.delete()
    try:
        replies = generate_replies(unliked)
    except Exception as e:
        await update.message.reply_text(f"Ошибка при генерации ответов: {e}")
        return

    for tweet, reply in zip(unliked, replies):
        await update.message.reply_text(
            _format_tweet_message(f"🐦 *Пост @{username}:*", username, tweet, reply),
            parse_mode="Markdown",
        )


async def get_posts_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not context.args:
        context.user_data["waiting_for_getposts"] = True
        await update.message.reply_text(
            "Напиши @username или ссылку на профиль X/Twitter:"
        )
        return

    username = parse_username(context.args[0])
    await _fetch_and_send_posts(update, context, username)


async def subscriptions_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await _show_subscriptions(update, update.effective_user.id)


async def subscription_callback(update: Update, context: ContextTypes.DEFAULT_TYPE):
    query = update.callback_query
    await query.answer()
    user_id = query.from_user.id

    if query.data == "sub_add":
        context.user_data["waiting_for_sub"] = True
        await query.message.reply_text(
            "Напиши @username или ссылку на аккаунт X/Twitter:"
        )
    elif query.data.startswith("sub_remove_"):
        username = query.data[len("sub_remove_"):]
        remove_sub(user_id, username)
        await _show_subscriptions(query, user_id, edit=True)


async def handle_text(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if context.user_data.get("waiting_for_getposts"):
        context.user_data["waiting_for_getposts"] = False
        username = parse_username(update.message.text)
        await _fetch_and_send_posts(update, context, username)
        return

    if context.user_data.get("waiting_for_sub"):
        context.user_data["waiting_for_sub"] = False
        username = parse_username(update.message.text)
        if not username:
            await update.message.reply_text("Не удалось распознать имя аккаунта.")
            return
        if add_sub(update.effective_user.id, username):
            await update.message.reply_text(f"✅ @{username} добавлен в подписки")
            await _check_one_subscription(update.effective_user.id, username, context.bot)
        else:
            await update.message.reply_text(f"@{username} уже в подписках")
        return

    await update.message.reply_text(
        "Используй:\n/getPosts @username — посты\n/subscriptions — подписки"
    )


async def _check_one_subscription(user_id: int, username: str, bot):
    try:
        unliked, _ = get_last_3_posts(username)
        new_unliked = [t for t in unliked if not was_sent(user_id, username, t["id"])]
        if not new_unliked:
            return
        replies = generate_replies(new_unliked)
        for tweet, reply in zip(new_unliked, replies):
            await bot.send_message(
                user_id,
                _format_tweet_message(f"🔔 *Новый пост @{username}:*", username, tweet, reply),
                parse_mode="Markdown",
            )
            mark_sent(user_id, username, tweet["id"])
    except Exception as e:
        logging.error(f"Subscription check failed for @{username} / user {user_id}: {e}")


async def _check_subscriptions(app: Application):
    data = load_subs()
    for user_id_str, usernames in data.items():
        user_id = int(user_id_str)
        for username in usernames:
            await _check_one_subscription(user_id, username, app.bot)


async def _subscription_loop(app: Application):
    while True:
        await asyncio.sleep(2 * 3600)
        logging.info("Running subscription check…")
        await _check_subscriptions(app)


async def main():
    token = os.getenv("TELEGRAM_BOT_TOKEN")
    if not token:
        raise RuntimeError("TELEGRAM_BOT_TOKEN не задан в .env")

    app = Application.builder().token(token).build()
    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("getPosts", get_posts_cmd))
    app.add_handler(CommandHandler("subscriptions", subscriptions_cmd))
    app.add_handler(CallbackQueryHandler(subscription_callback))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_text))

    me = await app.bot.get_me()
    print(f"Бот запущен: @{me.username} (id={me.id}). Ctrl+C для остановки.")

    async with app:
        await app.bot.delete_webhook(drop_pending_updates=False)
        await app.start()
        asyncio.create_task(_subscription_loop(app))
        await app.updater.start_polling()
        await asyncio.Event().wait()
        await app.updater.stop()
        await app.stop()


if __name__ == "__main__":
    asyncio.run(main())
